// fakebrowser/dist/cjs/BrowserBuilder.js
"use strict";
// noinspection JSUnusedGlobalSymbols
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BrowserBuilder = void 0;
const path = __importStar(require("path"));
const BrowserLauncher_1 = require("./BrowserLauncher");
const FakeBrowser_1 = require("./FakeBrowser");
class BrowserBuilder {
    constructor() {
        this.driverParams = {
            doNotHook: false,
            deviceDesc: FakeBrowser_1.kDefaultWindowsDD,
            userDataDir: '',
            evasionPaths: [
                'chrome.app',
                'chrome.csi',
                'chrome.loadTimes',
                'chrome.runtime',
                'window.history.length',
                'window.matchMedia',
                'navigator.webdriver',
                'sourceurl',
                'navigator.plugins-native',
                'webgl',
                'mimeTypes',
                'navigator.mediaDevices',
                'bluetooth',
                'navigator.permissions',
                'navigator.batteryManager',
                'webrtc',
                'canvas.fingerprint',
                'user-agent-override',
                'iframe.contentWindow',
                'iframe.src',
                'properties.getter',
                'font.fingerprint',
                'emoji.fingerprint',
                'window.speechSynthesis',
                'workers',
                'keyboard',
            ].map(e => path.resolve(__dirname, `../plugins/evasions/${e}`)),
            usePlugins: [],
        };
    }
    get launchParams() {
        const result = this.driverParams;
        result.launchOptions = result.launchOptions || {};
        return result;
    }
    get connectParams() {
        const result = this.driverParams;
        result.connectOptions = result.connectOptions || {};
        return result;
    }
    doNotHook(value) {
        this.launchParams.doNotHook = value;
        return this;
    }
    maxSurvivalTime(value) {
        this.launchParams.maxSurvivalTime = value;
        return this;
    }
    deviceDescriptor(value) {
        this.driverParams.deviceDesc = value;
        return this;
    }
    displayUserActionLayer(value) {
        this.driverParams.displayUserActionLayer = value;
        return this;
    }
    userDataDir(value) {
        this.driverParams.userDataDir = value;
        return this;
    }
    log(value) {
        this.driverParams.log = value;
        return this;
    }
    proxy(value) {
        this.driverParams.proxy = value;
        return this;
    }
    vanillaLaunchOptions(value) {
        this.launchParams.launchOptions = value;
        return this;
    }
    vanillaConnectOptions(value) {
        this.connectParams.connectOptions = value;
        return this;
    }
    evasionPaths(value) {
        this.driverParams.evasionPaths = value;
        return this;
    }
    usePlugins(value) {
        this.driverParams.usePlugins = value;
        return this;
    }
    async launch() {
        if ('undefined' === typeof this.launchParams.maxSurvivalTime) {
            this.launchParams.maxSurvivalTime = FakeBrowser_1.FakeBrowser.globalConfig.defaultBrowserMaxSurvivalTime;
        }
        const result = await BrowserLauncher_1.BrowserLauncher.launch(this.launchParams);
        return result;
    }
    async connect() {
        const result = await BrowserLauncher_1.BrowserLauncher.connect(this.connectParams);
        return result;
    }
}
exports.BrowserBuilder = BrowserBuilder;
