// fakebrowser/dist/cjs/Driver.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.kDefaultLaunchOptions = exports.kDefaultTimeout = void 0;
const assert_1 = require("assert");
const fs = __importStar(require("fs-extra"));
const puppeteer_extra_1 = require("puppeteer-extra");
const DeviceDescriptor_js_1 = __importDefault(require("./DeviceDescriptor.js"));
const UserAgentHelper_js_1 = require("./UserAgentHelper.js");
const PptrPatcher_1 = require("./PptrPatcher");
exports.kDefaultTimeout = 15 * 1000;
exports.kDefaultLaunchOptions = {
    headless: true,
    devtools: false,
    timeout: exports.kDefaultTimeout,
};
class Driver {
    static checkParamsLegal(params) {
        // deviceDesc must be set
        const dd = params.deviceDesc;
        (0, assert_1.strict)(dd, 'deviceDesc must be set');
        DeviceDescriptor_js_1.default.checkLegal(dd);
        // user data dir
        // The userDataDir in launchParameters must be set
        (0, assert_1.strict)(params.userDataDir, 'userDataDir must be set');
    }
    /**
     * Connect to browser
     * @param uuid
     * @param params
     */
    static async connect(uuid, params) {
        // Different instances with different puppeteer configurations
        const pptr = (0, puppeteer_extra_1.addExtra)(require('puppeteer'));
        // patch with evasions
        if (!params.doNotHook) {
            await PptrPatcher_1.PptrPatcher.patch(uuid, pptr, params);
        }
        const fakeDD = params.fakeDeviceDesc;
        (0, assert_1.strict)(!!fakeDD);
        const browser = await pptr.connect(params.connectOptions);
        await this.patchUAFromLaunchedBrowser(browser, fakeDD);
        return {
            vanillaBrowser: browser,
            pptrExtra: pptr,
        };
    }
    /**
     * Launch browser
     * @param uuid
     * @param defaultLaunchArgs
     * @param params
     */
    static async launch(uuid, defaultLaunchArgs, params) {
        this.checkParamsLegal(params);
        if (!params.launchOptions
            || Object.keys(params.launchOptions).length === 0) {
            params.launchOptions = exports.kDefaultLaunchOptions;
        }
        this.patchLaunchArgs(defaultLaunchArgs, params);
        // Different instances with different puppeteer configurations
        const pptr = (0, puppeteer_extra_1.addExtra)(require('puppeteer'));
        // patch with evasions
        if (!params.doNotHook) {
            await PptrPatcher_1.PptrPatcher.patch(uuid, pptr, params);
        }
        const fakeDD = params.fakeDeviceDesc;
        (0, assert_1.strict)(!!fakeDD);
        const browser = await pptr.launch(params.launchOptions);
        await this.patchUAFromLaunchedBrowser(browser, fakeDD);
        return {
            vanillaBrowser: browser,
            pptrExtra: pptr,
        };
    }
    static async patchUAFromLaunchedBrowser(browser, fakeDD) {
        // read major version from the launched browser and replace dd.userAgent
        const orgUA = await browser.userAgent();
        const orgVersion = UserAgentHelper_js_1.UserAgentHelper.chromeVersion(orgUA);
        const fakeVersion = UserAgentHelper_js_1.UserAgentHelper.chromeVersion(fakeDD.navigator.userAgent);
        (0, assert_1.strict)(orgVersion);
        (0, assert_1.strict)(fakeVersion);
        fakeDD.navigator.userAgent = fakeDD.navigator.userAgent.replace(fakeVersion, orgVersion);
        fakeDD.navigator.appVersion = fakeDD.navigator.appVersion.replace(fakeVersion, orgVersion);
    }
    static patchLaunchArgs(defaultLaunchArgs, launchParams) {
        // args
        // noinspection SuspiciousTypeOfGuard
        (0, assert_1.strict)(defaultLaunchArgs instanceof Array);
        const args = [
            ...defaultLaunchArgs,
            ...(launchParams.launchOptions.args || []),
        ];
        const fakeDD = launchParams.fakeDeviceDesc;
        (0, assert_1.strict)(!!fakeDD);
        // Modify default options
        launchParams.launchOptions = {
            ignoreHTTPSErrors: true,
            ignoreDefaultArgs: [
                '--enable-automation',
                '--enable-blink-features=IdleDetection',
            ],
            handleSIGINT: false,
            handleSIGTERM: false,
            handleSIGHUP: false,
            pipe: true,
            defaultViewport: {
                width: fakeDD.window.innerWidth,
                height: fakeDD.window.innerHeight,
                deviceScaleFactor: fakeDD.window.devicePixelRatio,
                isMobile: UserAgentHelper_js_1.UserAgentHelper.isMobile(fakeDD.navigator.userAgent),
                hasTouch: fakeDD.navigator.maxTouchPoints > 0,
                isLandscape: false,
            },
            ...launchParams.launchOptions,
            args,
        };
        // headless
        let headless = launchParams.launchOptions.headless;
        if ('undefined' === typeof headless) {
            headless = true;
        }
        if (launchParams.launchOptions.devtools) {
            headless = false;
        }
        // proxy
        if (launchParams.proxy) {
            args.push(`--proxy-server=${launchParams.proxy.proxy}`);
        }
        // browser language
        (0, assert_1.strict)(fakeDD.acceptLanguage);
        args.push(`--lang=${fakeDD.acceptLanguage}`);
        const userDataDir = launchParams.userDataDir;
        (0, assert_1.strict)(userDataDir);
        fs.mkdirSync(userDataDir, { recursive: true }); // throw exception
        args.push(`--user-data-dir=${userDataDir}`);
        // window position & window size
        let { screenX, screenY, innerWidth, innerHeight, outerWidth, outerHeight, } = fakeDD.window;
        outerWidth = outerWidth || innerWidth;
        outerHeight = outerHeight || (innerHeight + 85);
        args.push(`--window-position=${screenX},${screenY}`, `--window-size=${outerWidth},${outerHeight}`);
        // Some options can only be used in headless.
        // If you use them again in headful, you will see a plain white browser window without any content.
        if (headless) {
            args.push('--in-process-gpu', // https://source.chromium.org/search?q=lang:cpp+symbol:kInProcessGPU&ss=chromium
            '--disable-canvas-aa', // Disable antialiasing on 2d canvas
            '--disable-2d-canvas-clip-aa', // Disable antialiasing on 2d canvas clips
            '--disable-gl-drawing-for-tests');
        }
    }
    static async getPids(pid) {
        if ('string' === typeof (pid)) {
            pid = parseInt(pid);
        }
        try {
            const pidtree = require('pidtree');
            const pids = await pidtree(pid);
            return pids.includes(pid) ? pids : [...pids, pid];
        }
        catch (ignored) {
            return [pid];
        }
    }
    /**
     * Shutdown browser
     * @param browser
     */
    static async shutdown(browser) {
        try {
            const pages = await browser.pages();
            for (const page of pages) {
                await page.close();
            }
        }
        catch (ignored) {
        }
        const browserProcess = browser.process();
        if (browserProcess) {
            const pid = browserProcess.pid;
            if (pid) {
                const pids = await this.getPids(pid);
                pids.forEach(pid => {
                    try {
                        process.kill(pid, 'SIGKILL');
                    }
                    catch (ignored) {
                    }
                });
            }
        }
        try {
            await browser.close();
        }
        catch (ignored) {
        }
    }
}
exports.default = Driver;
