// fakebrowser/dist/cjs/PptrPatcher.js
"use strict";
// noinspection JSUnusedLocalSymbols,JSUnusedGlobalSymbols
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PptrPatcher = void 0;
const path = __importStar(require("path"));
const assert_1 = require("assert");
const axios_1 = __importDefault(require("axios"));
const helper_1 = require("./helper");
const FakeBrowser_1 = require("./FakeBrowser");
class PptrPatcher {
    static async patch(browserUUID, pptr, params) {
        (0, assert_1.strict)(!!params.fakeDeviceDesc);
        const opts = {
            browserUUID: browserUUID,
            internalHttpServerPort: FakeBrowser_1.FakeBrowser.globalConfig.internalHttpServerPort,
            fakeDD: params.fakeDeviceDesc,
            proxyExportIP: params.proxy && params.proxy.exportIP,
            myRealExportIP: await helper_1.helper.myRealExportIP(),
            historyLength: helper_1.helper.rd(2, 10),
        };
        // user action layer
        await this.patchUserActionLayer(browserUUID, pptr, params, opts);
        // evasions
        for (const evasionPath of params.evasionPaths) {
            const Plugin = require(evasionPath);
            const plugin = Plugin(opts);
            pptr.use(plugin);
        }
        // other plugins
        for (const plugin of params.usePlugins) {
            pptr.use(plugin);
        }
        // last, tidy up
        await this.patchLast(browserUUID, pptr, params, opts);
    }
    static patchUserActionLayer(uuid, pptr, params, opts) {
        if (params.displayUserActionLayer) {
            const Plugin = require(path.resolve(__dirname, '../plugins/user-action-layer'));
            const plugin = Plugin(opts);
            pptr.use(plugin);
        }
    }
    static patchLast(uuid, pptr, params, opts) {
        let Plugin = require(path.resolve(__dirname, '../plugins/evasions/zzzzzzzz.last'));
        let plugin = Plugin(opts);
        pptr.use(plugin);
    }
    /**
     * Package evasions to js string for worker to use
     * @param browser
     * @param jsContent
     */
    static async patchWorkerJsContent(browser, jsContent) {
        const jsPatch = await this.evasionsCode(browser);
        jsContent = jsPatch + jsContent;
        return jsContent;
    }
    static async evasionsCode(browser) {
        let jsPatch = '';
        const utils = require('../plugins/evasions/_utils');
        // utils
        let utilsContent = `const utils = {};\n`;
        for (const [key, value] of Object.entries(utils)) {
            const utilsFuncCode = value.toString();
            utilsContent += `utils.${key} = ${utilsFuncCode}; \n`;
        }
        utilsContent += `utils.init(); \n`;
        // code from puppeteer-extra
        const plugins = browser.pptrExtra.plugins;
        const runLast = plugins
            .filter(p => p.requirements.has('runLast'))
            .map(p => p.name);
        for (const name of runLast) {
            const index = plugins.findIndex(p => p.name === name);
            plugins.push(plugins.splice(index, 1)[0]);
        }
        for (const plugin of plugins) {
            if (plugin['onBrowser']) {
                await plugin['onBrowser'](browser.vanillaBrowser);
            }
            if (plugin['onServiceWorkerContent']) {
                // console.log(`SW Patch: ${plugin.name}`)
                jsPatch = await plugin['onServiceWorkerContent'](jsPatch);
            }
        }
        jsPatch = utilsContent + jsPatch;
        // when all evasions are patched, delete OffscreenCanvas.prototype.constructor.__cache
        return `(function() {${jsPatch};})(); \n\n 
const tmpVarNames =
    Object.getOwnPropertyNames(
        OffscreenCanvas.prototype.constructor,
    ).filter(e => e.startsWith('__$'));

tmpVarNames.forEach(e => {
    delete OffscreenCanvas.prototype.constructor[e];
});
`;
    }
    static async patchServiceWorkerRequest(browser, requestId, request, responseHeaders, client) {
        try {
            let base64Encoded = true;
            let jsContent;
            if (responseHeaders && responseHeaders.length) {
                let body;
                ({ body, base64Encoded } = await client.send('Fetch.getResponseBody', { requestId }));
                jsContent = base64Encoded ? Buffer.from(body, 'base64').toString('utf-8') : body;
            }
            else {
                // TODO: get through proxy
                const jsResp = await axios_1.default.get(request.url, { headers: request.headers });
                jsContent = jsResp.data;
                responseHeaders =
                    Object.entries(jsResp.headers).map(e => ({ name: e[0], value: e[1] }));
            }
            jsContent = await this.patchWorkerJsContent(browser, jsContent);
            // The order I used is: Fetch.enable -> on Fetch.requestPaused event -> Fetch.getResponseBody -> Fetch.fulfillRequest -> Fetch.disable
            await client.send('Fetch.fulfillRequest', {
                requestId,
                responseCode: 200,
                responseHeaders: responseHeaders,
                body: 1 ? Buffer.from(jsContent).toString('base64') : jsContent,
            });
            return true;
        }
        catch (ex) {
            console.error('SW inject failed', ex);
            await client.send('Fetch.failRequest', { requestId, errorReason: 'Aborted' });
        }
        return false;
    }
}
exports.PptrPatcher = PptrPatcher;
